langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java

```diff
@@ -0,0 +1,310 @@
+        this.retrievalQuery = getOrDefault(retrievalQuery, defaultRetrievalQuery);
+        
+        /* auto-schema creation */
+        createSchema();
+    }
+
+    /*
+    Methods with `@Override`
+    */
+    
+    @Override
+    public String add(Embedding embedding) {
+        String id = randomUUID();
+        add(id, embedding);
+        return id;
+    }
+
+    @Override
+    public void add(String id, Embedding embedding) {
+        addInternal(id, embedding, null);
+    }
+
+    @Override
+    public String add(Embedding embedding, TextSegment textSegment) {
+        String id = randomUUID();
+        addInternal(id, embedding, textSegment);
+        return id;
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings) {
+        return addAll(embeddings, null);
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+        List<String> ids = embeddings.stream()
+                .map(ignored -> randomUUID())
+                .toList();
+        addAllInternal(ids, embeddings, embedded);
+        return ids;
+    }
+
+    @Override
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
+        var embeddingValue = Values.value(referenceEmbedding.vector());
+
+        try (var session = session()) {
+            Map<String, Object> params = Map.of("indexName", indexName,
+                    "embeddingValue", embeddingValue,
+                    "minScore", minScore,
+                    "maxResults", maxResults);
+            return session
+                    .run("""
+						CALL db.index.vector.queryNodes($indexName, $maxResults, $embeddingValue)
+						YIELD node, score
+						WHERE score >= $minScore
+						""" + retrievalQuery, 
+                        params)
+                    .list(item -> Neo4jEmbeddingUtils.toEmbeddingMatch(this, item));
+        }
+    }
+
+    /*
+    Private methods
+    */
+    
+    private void addInternal(String id, Embedding embedding, TextSegment embedded) {
+        addAllInternal(singletonList(id), singletonList(embedding), embedded == null ? null : singletonList(embedded));
+    }
+
+    private void addAllInternal(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+        if (isCollectionEmpty(ids) || isCollectionEmpty(embeddings)) {
+            log.info("[do not add empty embeddings to neo4j]");
+            return;
+        }
+        ensureTrue(ids.size() == embeddings.size(), "ids size is not equal to embeddings size");
+        ensureTrue(embedded == null || embeddings.size() == embedded.size(), "embeddings size is not equal to embedded size");
+
+        bulk(ids, embeddings, embedded);
+    }
+
+    private void bulk(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+        Collection<List<Map<String, Object>>> rowsBatched = getRowsBatched(this, ids, embeddings, embedded);
```

Comment:
Would it be better to return a stream here instead of a collection? So that we stream one chunk at a time? And don't allocate all the memory upfront?